### PR TITLE
Initial query execution flow

### PR DESCRIPTION
**Pull request by Finistere**
# Description

Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.

# Type of change

- [ ] 💔 Breaking
- [ ] 🚀 Feature
- [ ] 🐛 Fix
- [ ] 🛠️ Tooling
- [ ] 🔨 Refactoring
- [ ] 🧪 Test
- [ ] 📦 Dependency
- [ ] 📖 Requires documentation update

### CODE REVIEW

File: engine/crates/engine-v2/src/response_graph/mod.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    Object(ObjectNodeId),
+}
+
+// Maybe an enum? Need to support `__typename`
+pub struct FieldEdge {
+    // probably needs a better name. it's the position for requested fields. For added fields,
+    // it's the position of the query field that needed it.
+    pub pos: Pos,
+    // resolving fragments eagerly, it makes manipulating SelectionSet easier during planning.
+    pub type_condition: Option<TypeCondition>,
```
Comment (obmarg) [inline engine/crates/engine-v2/src/response_graph/mod.rs:120]: Have you given any thought to the other uses of fragments?  I can see this approach working alright for type conditions, but not so sure about `@defer`, `@skip`, `@include`?  Not so sure

File: engine/crates/engine-v2/src/plan/logical/graph.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
+use super::LogicalPlan;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Hash)]
+pub struct PlanId(usize); // should be a u16, more than enough for any sane query
```
Comment (obmarg) [inline engine/crates/engine-v2/src/plan/logical/graph.rs:4]: I'd vote u32 but sure

File: engine/crates/engine-v2/src/response_graph/mod.rs (hunk 1 above)
Comment (Finistere) [reply engine/crates/engine-v2/src/response_graph/mod.rs]: excellent point. for `@skip` & `@include` it shouldn't be hard as the condition can be detected quite easily. For `@defer`, it's definitely more tricky. I think in this case we would probably have a `deferred` field in the `NodeSelectionSet` containing the deferred fragment fields. A data source might only be responsible of only some parts of fragment fields, so I'm a bit skeptical of keeping the fragment notion which does add more complexity. The only benefit I see is more readable queries generated against subgraphs. A valid one, but not critical imho.

File: engine/crates/engine-v2/src/plan/logical/graph.rs (hunk 2 above)
Comment (Finistere) [reply engine/crates/engine-v2/src/plan/logical/graph.rs]: removed the comment, we'll see it another day. Given how the plans are currently generated I think 65K is more than enough. But it's a decision that can be easily changed later. 🤷

File: engine/crates/engine-v2/src/executor/mod.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+impl Executor {
+    fn build(engine: &Engine, resolver: &schema::Resolver, request: ExecutorRequest<'_>) -> Self {
+        match resolver {
+            Resolver::Subgraph(resolver) => graphql::GraphqlExecutor::build(engine, resolver, request),
+        }
+    }
+
+    async fn execute(self, proxy: ResponseGraphProxy) {
+        match self {
+            Executor::GraphQL(plan) => plan.execute(proxy).await,
```
Comment (obmarg) [inline engine/crates/engine-v2/src/executor/mod.rs:36]: Nitpicking , but the thing inside the `Executor::GraphQL` is a `GraphQlExecutor` - so `executor` seems like a less confusing name.

File: engine/crates/engine-v2/schema/src/conversion.rs [hunk 4]
```diff
@@ -281,3 +299,9 @@ from_id_newtypes! {
     SubgraphId => SubgraphId,
     UnionId => UnionId,
 }
+
+impl DataSourceId {
+    fn from_subgraph_id(id: federated_graph::SubgraphId) -> Self {
+        Self(NonZeroU32::new(id.0 as u32 + 1).unwrap())
```
Comment (obmarg) [inline engine/crates/engine-v2/schema/src/conversion.rs:305]: Not sure I'm _against_ these `NonZeroU32`s - but they do seem like they're increasing our risk of off by one errors significantly...

**Review by obmarg — COMMENTED** [review]
Not finished reviewing yet, but half my requests are for docstrings so going to submit my review and come back to it later

File: engine/crates/engine-v2/schema/src/lib.rs [hunk 5]
```diff
@@ -76,8 +87,88 @@ pub struct Key {
     pub fields: SelectionSet,
 }
 
-pub type SelectionSet = Vec<Selection>;
+#[derive(Default, Clone)]
+pub struct SelectionSet {
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/schema/src/lib.rs:87]: The schema contains selection sets for `@requires` & `@provides`

File: engine/crates/engine-v2/schema/src/lib.rs [hunk 6]
```diff
@@ -76,8 +87,88 @@ pub struct Key {
...
+    // sorted by field id
+    items: Vec<Selection>,
+}
+
+impl FromIterator<Selection> for SelectionSet {
+    fn from_iter<T: IntoIterator<Item = Selection>>(iter: T) -> Self {
+        let mut items = iter.into_iter().collect::<Vec<_>>();
+        items.sort_unstable_by_key(|selection| selection.field);
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/schema/src/lib.rs:95]: Mostly to make `merge` easier as I'm manipulating those selection sets directly during planning. It's true that the spec doesn't explicitly specify those as unordered. But I don't think it really matters, no? In this case, it's the selection set used to represent `@provides` and `@requires`, but I'm ignoring any user order everywhere else for the response.

File: engine/crates/engine-v2/src/executor/mod.rs (hunk 3 above)
Comment (Finistere) [reply engine/crates/engine-v2/src/executor/mod.rs]: yeah leftover

File: engine/crates/engine-v2/schema/src/conversion.rs (hunk 4 above)
Comment (Finistere) [reply engine/crates/engine-v2/schema/src/conversion.rs]: Very fair point, moved all the IDs into a sub-module so the only way to create them is through `From<usize>`. As I'm using `Option<ObjectId>` in response objects it felt like a waste of space to allocate additional bytes.

File: engine/crates/engine-v2/schema/src/translator.rs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+use crate::{FieldId, InterfaceId, ObjectId, Schema, UnionId};
+
+const TYPENAME: &str = "__typename";
+
+pub trait Translator {
+    fn field(&self, field_id: FieldId) -> &str;
+    fn typename(&self, object_id: ObjectId) -> &str;
+    fn union_typename_field(&self, union_id: UnionId) -> &str;
+    fn interface_typename_field(&self, interface_id: InterfaceId) -> &str;
+}
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/schema/src/translator.rs]: removed it. The idea was to allow resolvers to define how to rename/translate fields. It was simple when reading from the response but a lot more tricky when writing data into it.

File: engine/crates/engine-v2/src/request.rs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                            schema::Definition::Object(object_id) => self.schema.object_fields(object_id).collect(),
+                            schema::Definition::Interface(interface_id) => {
+                                self.schema.interface_fields(interface_id).collect()
+                            }
+                            _ => {
+                                return Err(ServerError::new(
+                                    format!("Field {name} does not have any fields."),
+                                    Some(pos),
+                                ));
+                            }
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/request.rs]: yeah indeed, good catch

File: engine/crates/engine-v2/src/response_graph/selection_set/output.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
+use super::{NodeSelection, NodeSelectionSet};
+use crate::response_graph::FieldEdgeId;
+
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct OutputNodeSelectionSet {
+    items: Vec<OutputNodeSelection>,
+}
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/response_graph/selection_set/output.rs]: renamed those to `ReadSelectionSet` and `WriteSelectionSet`. They're respectively used to read from the response and write into it. So plan inputs and outputs. Compared to the bare `SelectionSet` they had more attributes to rename fields on the fly. Currently, they're the same after removing code, but it makes sense to me to keep them nonetheless imho.

File: engine/crates/engine-v2/src/response_graph/input.rs [hunk 10]
```diff
@@ -0,0 +1,114 @@
+use serde::ser::{SerializeMap, SerializeSeq};
+use serde_json::Value;
+
+use super::{InputNodeSelectionSet, Node, ObjectNode, ObjectNodeId, ResponseGraph};
+
+pub struct Input<'a> {
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/response_graph/input.rs]: renamed to `ResponseObjectsView` it's used as the input for a plan avoiding any copies

File: engine/crates/engine-v2/schema/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (obmarg) [reply engine/crates/engine-v2/schema/src/lib.rs:97]: The spec explicitly specifies that the order of output should match the order of the query: http://spec.graphql.org/October2021/#sec-Serialized-Map-Ordering

[the remark this answers is not in the document]
Comment (obmarg) [reply engine/crates/engine-v2/schema/src/lib.rs:97]: Mutation fields also need to be executed serially in the order that they were specified in the operation: http://spec.graphql.org/October2021/#sec-Mutation

[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/schema/src/lib.rs:97]: Excellent catch.

File: engine/crates/engine-v2/src/executor/coordinator.rs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+                        // We should data back through a stream to later support additional data
+                        // being sent like @defer or streaming
+                        break;
+                    }
+                }
+            }
+        }
+
+        // Response will be sent through a byte stream back. Client shouldn't be aware of wether we
+        // defer / stream / etc.
```
Comment (obmarg) [inline engine/crates/engine-v2/src/executor/coordinator.rs:85]: > Client shouldn't be aware of whether we // defer / stream / etc.

Just a note for the future: I don't think this layer should deal with the byte-stream aspect of it.  This should probably return a stream of some `Response: Serialize` type, and leave the converting-to-byte-stream bit up to a higher layer.

Comment (Finistere) [reply engine/crates/engine-v2/src/executor/coordinator.rs:85]: indeed, would definitely be better

Comment (obmarg) [reply engine/crates/engine-v2/src/executor/coordinator.rs:85]: There's far too many different transports (and http libraries) for the core engine to be concerning itself with all those details

File: engine/crates/engine-v2/src/request.rs [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+            OperationType::Mutation => self
+                .schema
+                .root_operation_types
+                .mutation
+                .expect("Mutation operation type not supported by schema."),
+            OperationType::Subscription => self
+                .schema
+                .root_operation_types
+                .subscription
+                .expect("Subscription operation type not supported by schema."),
```
Comment (obmarg) [inline engine/crates/engine-v2/src/request.rs:42]: This function returns a Result, so these should probably be `ok_or_else(...)?`

**Review by obmarg — APPROVED** [review]
Good work.  Still a bunch to do, but lets get this merged sometime and iterate

File: engine/crates/engine-v2/src/request.rs (hunk 12 above)
Comment (Finistere) [reply engine/crates/engine-v2/src/request.rs:42]: oh yeah definitely. I'll need to make a first pass over the code to add some error management which is pretty much nonexistent currently. ^^

**Pull request merged**